packages/plugins/other/visitor-plugin-common/src/client-side-base-visitor.ts

```diff
@@ -63,21 +71,61 @@ export interface ClientSideBasePluginConfig extends ParsedConfig {
    */
   dedupeOperationSuffix: boolean;
   noExport: boolean;
+
+  /**
+   * @name documentMode
+   * @type 'graphQLTag' | 'documentNode' | 'external'
+   * @default 'graphQLTag'
+   * @description Declares how DocumentNode are created:
+   * - `graphQLTag`: `graphql-tag` or other modules (check `gqlImport`) will be used to generate document nodes. If this is used, document nodes are generated on client side i.e. the module used to generate this will be shipped to the client
+   * - `documentNode`: document nodes will be generated as objects when we generate the templates.
+   * - `external`: document nodes are imported from an external file. To be used with `importDocumentNodeExternallyFrom`
+   */
+  documentMode?: DocumentMode;
+
+  /**
+   * @name importDocumentNodeExternallyFrom
+   * @type string | 'near-operation-file'
+   * @default ''
+   * @description This config should be used if `documentMode` is `external`. This has 2 usage:
+   * - any string: This would be the path to import document nodes from. This can be used if we want to manually create the document nodes e.g. Use `graphql-tag` in a separate file and export the generated document
+   * - 'near-operation-file': This is a special mode that is intended to be used with `near-operation-file` preset to import document nodes from those files. If these files are `.graphql` files, we make use of webpack loader.
+   *
+   * @example
+   * ```yml
+   * config:
+   *   documentMode: external
+   *   importDocumentNodeExternallyFrom: path/to/document-node-file
+   * ```
+   *
+   * ```yml
+   * config:
+   *   documentMode: external
+   *   importDocumentNodeExternallyFrom: near-operation-file
+   * ```
+   *
+   */
+  importDocumentNodeExternallyFrom?: string;
 }
 
 export class ClientSideBaseVisitor<TRawConfig extends RawClientSideBasePluginConfig = RawClientSideBasePluginConfig, TPluginConfig extends ClientSideBasePluginConfig = ClientSideBasePluginConfig> extends BaseVisitor<TRawConfig, TPluginConfig> {
   protected _collectedOperations: OperationDefinitionNode[] = [];
+  protected _documents: Types.DocumentFile[] = [];
 
-  constructor(protected _fragments: LoadedFragment[], rawConfig: TRawConfig, additionalConfig: Partial<TPluginConfig>) {
+  constructor(protected _fragments: LoadedFragment[], rawConfig: TRawConfig, additionalConfig: Partial<TPluginConfig>, documents?: Types.DocumentFile[]) {
     super(rawConfig, {
       dedupeOperationSuffix: getConfigValue(rawConfig.dedupeOperationSuffix, false),
       noGraphQLTag: getConfigValue(rawConfig.noGraphQLTag, false),
```

Comment:
I think you can remove `noGraphQLTag` from the parsed config, and just do? 
```
documentMode: typeof rawConfig.noGraphQLTag === 'boolean' && typeof rawConfig.noGraphQLTag === false ? 'documentMode' : getConfigValue(rawConfig.documentMode, 'graphQLTag'),
importDocumentNodeExternallyFrom: getConfigValue(rawConfig.importDocumentNodeExternallyFrom, ''),
```

Reply:
Interesting! I've always wondered why we had raw and parsed config. I have fixed this in https://github.com/dotansimha/graphql-code-generator/pull/2261/commits/42ffe06eb4adaa14050a1a5590ecfa2437b5cdbf